File: kindlife_app/kindlife_app/report/pending_item_price_approval/pending_item_price_approval.js
frappe.query_reports["Pending Item Price Approval"] = {
    "filters": [
        {
            "fieldname": "price_list",
            "label": __("Price List"),
            "fieldtype": "Link",
            "options": "Price List"
        },
        {
            "fieldname": "item_code",
            "label": __("Item"),
            "fieldtype": "Link",
            "options": "Item"
        }
    ],
    
    "onload": function(report) {
        // Add bulk action buttons
        report.page.add_inner_button(__('Bulk Approve Selected'), function() {
            bulk_action('approve',report);
        }, __('Actions'));
        
        report.page.add_inner_button(__('Bulk Reject Selected'), function() {
            bulk_action('reject',report);
        }, __('Actions'));
        
        // report.page.add_inner_button(__('Select All'), function() {
        //     toggle_all_checkboxes(true);
        // }, __('Selection'));
        
        // report.page.add_inner_button(__('Deselect All'), function() {
        //     toggle_all_checkboxes(false);
        // }, __('Selection'));
        
        // Add custom CSS for better checkbox styling
        $('<style>')
            .text(`
                .row-check {
                    transform: scale(1.2);
                    margin: 0;
                }
                .bulk-actions-section {
                    padding: 10px;
                    background: #f8f9fa;
                    border-bottom: 1px solid #dee2e6;
                    margin-bottom: 10px;
                }
                .selected-count {
                    font-weight: bold;
                    color: #007bff;
                }
            `)
            .appendTo('head');
    },
	get_datatable_options(options) {
        return Object.assign(options, {
            checkboxColumn: true,
        });
    },
    
    "formatter": function(value, row, column, data, default_formatter) {
        if (column.fieldname == "actions") {
            return `
                <div style="display: flex; gap: 3px; justify-content: center;">
                    <button class='btn btn-xs btn-success' 
                            style='padding: 2px 6px; font-size: 10px;'
                            onclick='approvePrice("${data.name}")'
                            title='Approve'>
                        <i class="fa fa-check"></i>
                    </button>
                    <button class='btn btn-xs btn-danger' 
                            style='padding: 2px 6px; font-size: 10px;'
                            onclick='rejectPrice("${data.name}")'
                            title='Reject'>
                        <i class="fa fa-times"></i>
                    </button>
                </div>
            `;
        }
        
        return default_formatter(value, row, column, data);
    }
};

// Global functions for individual actions
window.approvePrice = function(item_price_name) {
    frappe.confirm(
        __('Are you sure you want to approve this price change?'),
        function() {
            frappe.call({
                method: 'kindlife_app.api.item_price.bulk_approve_prices',
                args: {
                    'item_price_names': [item_price_name]
                },
                callback: function(r) {
                    if (r.message && r.message.success) {
                        frappe.show_alert({
                            message: __('Price approved successfully'),
                            indicator: 'green'
                        });
                        frappe.query_report.refresh();
                    } else {
                        frappe.msgprint({
                            title: __('Error'),
                            message: r.message ? r.message.message : __('Failed to approve price'),
                            indicator: 'red'
                        });
                    }
                }
            });
        }
    );
};

window.rejectPrice = function(item_price_name) {
    frappe.confirm(
        __('Are you sure you want to reject this price change?'),
        function() {
            frappe.call({
                method: 'kindlife_app.api.item_price.bulk_reject_prices',
                args: {
                    'item_price_names': [item_price_name]
                },
                callback: function(r) {
                    if (r.message && r.message.success) {
                        frappe.show_alert({
                            message: __('Price rejected successfully'),
                            indicator: 'orange'
                        });
                        frappe.query_report.refresh();
                    } else {
                        frappe.msgprint({
                            title: __('Error'),
                            message: r.message ? r.message.message : __('Failed to reject price'),
                            indicator: 'red'
                        });
                    }
                }
            });
        }
    );
};

// Bulk action functions
function bulk_action(action, report) {
    let selected_items = get_selected_items(report);
    
    if (selected_items.length === 0) {
        frappe.msgprint(__('Please select at least one item'));
        return;
    }
    
    let action_text = action === 'approve' ? 'approve' : 'reject';
    let method = action === 'approve' ? 'bulk_approve_prices' : 'bulk_reject_prices';
    
    frappe.confirm(
        __(`Are you sure you want to ${action_text} ${selected_items.length} selected item(s)?`),
        function() {
            frappe.call({
                method: `kindlife_app.api.item_price.${method}`,
                args: {
                    'item_price_names': selected_items
                },
                callback: function(r) {
                    if (r.message && r.message.success) {
                        let count = action === 'approve' ? r.message.approved_count : r.message.rejected_count;
                        frappe.show_alert({
                            message: __(`Successfully ${action_text}ed ${count} item(s)`),
                            indicator: action === 'approve' ? 'green' : 'orange'
                        });
                        
                        if (r.message.failed_items && r.message.failed_items.length > 0) {
                            frappe.msgprint({
                                title: __('Partial Success'),
                                message: __('Some items failed to process:') + '<br>' + 
                                        r.message.failed_items.join('<br>'),
                                indicator: 'yellow'
                            });
                        }
                        
                        report.refresh();
                    } else {
                        frappe.msgprint({
                            title: __('Error'),
                            message: r.message ? r.message.message : __(`Failed to ${action_text} selected items`),
                            indicator: 'red'
                        });
                    }
                }
            });
        }
    );
}

function get_selected_items(report) {
    let selected_items = [];
    
    if (report && report.datatable) {
        // Get selected row indices
        let selected_indexes = report.datatable.rowmanager.getCheckedRows();
        
        // Get the actual data for selected rows
        selected_indexes.forEach(function(index) {
            if (report.data && report.data[index]) {
                selected_items.push(report.data[index].name);
            }
        });
    }
    
    return selected_items;
}

function toggle_all_checkboxes(select) {
    $('.row-check').prop('checked', select);
    updateSelectionCount();
}

function updateSelectionCount() {
    let selected_count = $('.row-check:checked').length;
    let total_count = $('.row-check').length;
    
    // Update selection info in the report header
    let selection_info = $(`.selection-info`);
    if (selection_info.length === 0) {
        $('.report-wrapper .page-head').after(`
            <div class="bulk-actions-section">
                <span class="selection-info">
                    Selected: <span class="selected-count">0</span> of <span class="total-count">${total_count}</span>
                </span>
            </div>
        `);
        selection_info = $('.selection-info');
    }
    
    selection_info.find('.selected-count').text(selected_count);
    selection_info.find('.total-count').text(total_count);
    
    // Show/hide bulk action buttons based on selection
    if (selected_count > 0) {
        $('.bulk-actions-section').addClass('has-selection');
    } else {
        $('.bulk-actions-section').removeClass('has-selection');
    }
}

File: kindlife_app/kindlife_app/report/pending_item_price_approval/pending_item_price_approval.py
import frappe
from frappe import _

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_pending_columns()
    data = get_pending_data(filters)
    
    return columns, data

def get_pending_columns():
    return [
        #  {
        #     "fieldname": "select_row",
        #     "label": "",
        #     "fieldtype": "Data",
        #     "width": 25
        # },
		{
            "label": _("Item Price"),
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Item Price",
            "width": 100
        },
        {
            "label": _("Item Code"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120
        },
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("Price List"),
            "fieldname": "price_list",
            "fieldtype": "Link",
            "options": "Price List",
            "width": 150
        },
        {
            "label": _("MRP(Cur.)"),
            "fieldname": "current_buying_price_formatted",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Buying Price(Cur.)"),
            "fieldname": "current_price_list_rate_formatted",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("MRP(Req.)"),
            "fieldname": "custom_buying_price_formatted",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("Buying Price(Req.)"),
            "fieldname": "price_list_rate_formatted",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Requested By"),
            "fieldname": "custom_requested_by",
            "fieldtype": "Link",
            "options": "User",
            "width": 120
        },
        {
            "label": _("Request Date"),
            "fieldname": "creation",
            "fieldtype": "Datetime",
            "width": 140
        },
        {
            "label": _("Valid From"),
            "fieldname": "valid_from",
            "fieldtype": "Date",
            "width": 100
        },
        {
            "label": _("Actions"),
            "fieldname": "actions",
            "fieldtype": "Data",
            "width": 150
        }
    ]

def get_pending_data(filters):
	conditions = [ "ip.workflow_state = 'Pending'", "ip.custom_disable = 0"]

	if filters.get('price_list'):
		conditions.append("ip.price_list = %(price_list)s")

	if filters.get('item_code'):
		conditions.append("ip.item_code = %(item_code)s")

	query = f"""
				SELECT 
					ip.name,
					ip.item_code,
					i.item_name,
					ip.price_list,
					ip.price_list_rate,
					ip.custom_buying_price,
					ip.custom_requested_by,
					ip.creation,
					ip.valid_from,
					ip_cur.price_list_rate AS current_price_list_rate,
					ip_cur.custom_buying_price AS current_buying_price,
					ip.currency
				FROM 
					`tabItem Price` ip
				LEFT JOIN 
					`tabItem` i ON ip.item_code = i.name
				
				LEFT JOIN (
					SELECT 
						ip2.item_code,
						ip2.price_list,
						ip2.price_list_rate,
						ip2.custom_buying_price
					FROM 
						`tabItem Price` ip2
					WHERE 
						ip2.workflow_state = 'Approved'
						AND ip2.custom_disable = 0
						AND ip2.valid_from = (
							SELECT MAX(ip3.valid_from)
							FROM `tabItem Price` ip3
							WHERE 
								ip3.item_code = ip2.item_code
								AND ip3.price_list = ip2.price_list
								AND ip3.valid_from <= CURDATE()
								AND ip3.workflow_state = 'Approved'
								AND ip3.custom_disable = 0
						)
						AND ip2.name = (
						SELECT ip4.name
						FROM `tabItem Price` ip4
						WHERE ip4.item_code = ip2.item_code
						AND ip4.price_list = ip2.price_list
						AND ip4.workflow_state = 'Approved'
						AND ip4.valid_from = ip2.valid_from
						AND ip4.custom_disable = 0
						ORDER BY ip4.modified DESC
						LIMIT 1
					)
				) AS ip_cur 
				ON ip_cur.item_code = ip.item_code AND ip_cur.price_list = ip.price_list
				WHERE 
					{' AND '.join(conditions)}
				ORDER BY 
					ip.creation DESC
			"""

	data = frappe.db.sql(query, filters, as_dict=1)

	print("data",data)

	# Format currency values and add action buttons
	for row in data:
		# Format currency values with proper currency symbol
		if row.get('currency'):
			currency_symbol = frappe.db.get_value('Currency', row.get('currency'), 'symbol') or row.get('currency')
			
			# Format current buying price
			if row.get('current_buying_price'):
				row['current_buying_price_formatted'] = f"{frappe.utils.fmt_money(row['current_buying_price'], currency=row.get('currency'))}"
			else:
				row['current_buying_price_formatted'] = f"{currency_symbol} 0.00"
				
			# Format current price list rate
			if row.get('current_price_list_rate'):
				row['current_price_list_rate_formatted'] = f"{frappe.utils.fmt_money(row['current_price_list_rate'], currency=row.get('currency'))}"
			else:
				row['current_price_list_rate_formatted'] = f"{currency_symbol} 0.00"
				
			# Format requested buying price
			if row.get('custom_buying_price'):
				row['custom_buying_price_formatted'] = f"{frappe.utils.fmt_money(row['custom_buying_price'], currency=row.get('currency'))}"
			else:
				row['custom_buying_price_formatted'] = f"{currency_symbol} 0.00"
				
			# Format requested price list rate
			if row.get('price_list_rate'):
				row['price_list_rate_formatted'] = f"{frappe.utils.fmt_money(row['price_list_rate'], currency=row.get('currency'))}"
			else:
				row['price_list_rate_formatted'] = f"{currency_symbol} 0.00"
		else:
			# Fallback if currency is not available
			row['current_buying_price_formatted'] = str(row.get('current_buying_price', 0))
			row['current_price_list_rate_formatted'] = str(row.get('current_price_list_rate', 0))
			row['custom_buying_price_formatted'] = str(row.get('custom_buying_price', 0))
			row['price_list_rate_formatted'] = str(row.get('price_list_rate', 0))

		# Add action buttons
		row['actions'] = f"""
			<button class='btn btn-xs btn-success' 
					onclick='approvePrice("{row.name}")'>
				Approve
			</button>
			<button class='btn btn-xs btn-danger' 
					onclick='rejectPrice("{row.name}")'>
				Reject
			</button>
		"""

	return data